packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/TaskManagerInternalModule.java

```diff
@@ -0,0 +1,142 @@
+
+  @Override
+  public void setModuleRegistry(ModuleRegistry moduleRegistry) {
+    mEventEmitter = moduleRegistry.getModule(EventEmitter.class);
+    mConstants = moduleRegistry.getModule(ConstantsInterface.class);
+    mTaskService = moduleRegistry.getSingletonModule("TaskService", TaskServiceInterface.class);
+
+    // Register in TaskService.
+    mTaskService.setTaskManager(this, getAppId(), getAppUrl());
+  }
+
+  //endregion
+  //region TaskManagerInterface
+
+  @Override
+  public void registerTask(String taskName, Class consumerClass, Map<String, Object> options) throws Exception {
+    checkTaskService();
+    mTaskService.registerTask(taskName, getAppId(), getAppUrl(), consumerClass, options);
+  }
+
+  @Override
+  public void unregisterTask(String taskName, Class consumerClass) throws Exception {
+    checkTaskService();
+    mTaskService.unregisterTask(taskName, getAppId(), consumerClass);
+  }
+
+  @Override
+  public void executeTaskWithBody(Bundle body) {
+    if (mEventsQueue != null) {
+      // `startObserving` on TaskManagerModule wasn't called yet - add event body to the queue.
+      mEventsQueue.add(body);
+    } else {
+      // Manager is already being observed by JS app, so we can execute the event immediately.
+      mEventEmitter.emit(TaskManagerModule.EVENT_NAME, body);
+    }
+  }
+
+  @Override
+  public boolean taskHasConsumerOfClass(String taskName, Class consumerClass) {
+    if (mTaskService == null) {
+      return false;
+    }
+    return mTaskService.taskHasConsumerOfClass(taskName, getAppId(), consumerClass);
+  }
+
+  @Override
+  public void flushQueuedEvents() {
+    // Execute any events that came before this call.
+    if (mEventsQueue != null) {
+      for (Bundle body : mEventsQueue) {
+        mEventEmitter.emit(TaskManagerModule.EVENT_NAME, body);
+      }
+      mEventsQueue = null;
+    }
+  }
+
+  @Override
+  public String getAppId() {
+    if (mConstants != null) {
+      return mConstants.getAppId();
+    }
+    return null;
+  }
+
+  @Override
+  public boolean isRunningInHeadlessMode() {
+    if (mConstants != null) {
+      return (boolean) mConstants.getConstants().get("isHeadless");
+    }
+    return false;
+  }
+
+  //endregion
+  //region helpers
+
+  private String getAppUrl() {
+    // If Constants module is available and experienceUrl is provided, just return it.
+    if (mConstants != null) {
+      String experienceUrl = (String) mConstants.getConstants().get("experienceUrl");
+      if (experienceUrl != null) {
+        return experienceUrl;
+      }
+    }
+
+    // Fallback to package name? It should work well for RN apps with just one JS app inside.
+    Context context = mContextRef.get();
+    if (context != null) {
+      return context.getPackageName();
+    }
+
+    return null;
+  }
+
+  private void checkTaskService() throws Exception {
+    if (mTaskService == null) {
+      throw new Exception("Unable to find TaskService singleton module in module registry.");
```

Comment:
```suggestion
      throw new IllegalStateException("Unable to find TaskService singleton module in module registry.");
```
and then the `throws` clause?